New resources: `azurerm_iothub_device_update_account` `azurerm_iothub_device_update_instance`

myc2h6o opened the pull request:
- Adding the support for these two new resources for IoT Hub Resource Management: https://learn.microsoft.com/azure/iot-hub-device-update/create-device-update-account
- Swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/deviceupdate/resource-manager/Microsoft.DeviceUpdate/stable/2022-10-01/deviceupdate.json
- These two resources are part of IoT Hub Management, so I put them under the `iothub` directory
- IoT Hub Device Update is not available in `westeurope`, so updating the teamcity locations with `northeurope` instead

stephybun reviewed (commented):
Thanks for this PR @myc2h6o, there are still some test failures relating to the region though
```
------- Stdout: -------
=== RUN   TestAccIotHubDeviceUpdateAccount_sku
=== PAUSE TestAccIotHubDeviceUpdateAccount_sku
=== CONT  TestAccIotHubDeviceUpdateAccount_sku
testcase.go:110: Step 1/4 error: Error running apply: exit status 1
Error: creating Account (Subscription: "*******"
Resource Group Name: "acctest-rg-221018105259715989"
Account Name: "acc-dua-eu3sy"): performing AccountsCreate: deviceupdates.DeviceupdatesClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="LocationNotAvailableForResourceType" Message="The provided location 'westeurope' is not available for resource type 'Microsoft.DeviceUpdate/accounts'. List of available regions for the resource type is 'westus2,northeurope,southeastasia,eastus,eastus2,southcentralus,swedencentral,uksouth,westus3'."
with azurerm_iothub_device_update_account.test,
on terraform_plugin_test.tf line 13, in resource "azurerm_iothub_device_update_account" "test":
13: resource "azurerm_iothub_device_update_account" "test" {
creating Account (Subscription: "*******"
Resource Group Name: "acctest-rg-221018105259715989"
Account Name: "acc-dua-eu3sy"): performing AccountsCreate:
deviceupdates.DeviceupdatesClient#AccountsCreate: Failure sending request:
StatusCode=0 -- Original Error: Code="LocationNotAvailableForResourceType"
Message="The provided location 'westeurope' is not available for resource
type 'Microsoft.DeviceUpdate/accounts'. List of available regions for the
resource type is
'westus2,northeurope,southeastasia,eastus,eastus2,southcentralus,swedencentral,uksouth,westus3'."
--- FAIL: TestAccIotHubDeviceUpdateAccount_sku (120.85s)
FAIL
```

myc2h6o commented:
Hi @stephybun thanks for reviewing the pr! I've already updated the region here: https://github.com/hashicorp/terraform-provider-azurerm/pull/18789/files#diff-c9a4f5ece9fe921f274cae0220ba39a13027c503c6bd80e3cc9ee926565a5c94, to replace the `westeurope` with `northeurope`, however the test seems to be run within westeurope still. Does it require the change to be merged to allow the config to take effect or am I doing something wrong there?

The test runs well with overriden parameter:
![image](https://user-images.githubusercontent.com/10579712/196652685-788e4181-4dd6-4f31-9841-fc0745ac6b67.png)

myc2h6o commented:
The separate pr for updating test location is working fine. Here is the updated test result
![image](https://user-images.githubusercontent.com/10579712/197241573-f41a81fd-10d9-443c-87e8-ad337cedf40e.png)

jackofallops commented on internal/services/iothub/iothub_device_update_account_resource.go:
```diff
@@ -0,0 +1,284 @@
+				existing.Tags = &model.Tags
+			}
+
+			existing.Properties.HostName = nil
+			existing.Properties.Locations = nil
+			existing.SystemData = nil
```
This shouldn't need to be zeroed, ARM spec says the service is supposed to ignore R/O fields on update payloads? Needing this would indicated a bug?

jackofallops commented on internal/services/iothub/iothub_device_update_instance_resource.go:
```diff
@@ -0,0 +1,325 @@
+				existing.Tags = &model.Tags
+			}
+
+			existing.SystemData = nil
```
As above, this shouldn't be necessary.

jackofallops commented on internal/services/iothub/validate/iot_hub_device_update_instance_name.go:
```diff
@@ -0,0 +1,20 @@
+	}
+
+	if matched := regexp.MustCompile(`^[A-Za-z0-9]+(-[A-Za-z0-9]+)*$`).Match([]byte(value)); !matched {
+		errors = append(errors, fmt.Errorf("%q may only contain alphanumeric characters and dashes, and consecutive dashes (-) are not allowed", k))
```
```suggestion
		errors = append(errors, fmt.Errorf("%q must start with an alphanumeric, may only contain alphanumeric characters and dashes, and consecutive dashes (-) are not allowed", k))
```

myc2h6o replied:
Error message updated.

jackofallops reviewed (changes requested):
Thanks @myc2h6o 
The update methods I think need revisiting. The removal of the data from the payloads there is an anti-pattern that shouldn't be necessary and we should avoid within the provider. Can you speak with the service team on your side to resolve?

Thanks!

myc2h6o commented:
Hi @jackofallops thanks for reviewing the pr! I've updated the error message, and removed the unexpected assignment of `HostName` and `Locations` in the Update function.
As for `SystemData`, it is a workaround before the Pandora side fix https://github.com/hashicorp/pandora/issues/1096 is done, similar to [signalr_service](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/signalr/signalr_service_network_acl_resource.go#L200-L201), I've added comment around the code.

myc2h6o commented:
Test result after the update
![image](https://user-images.githubusercontent.com/10579712/197666435-a28acfbf-9206-4a07-a341-554be8efa161.png)

katbyte commented on website/docs/r/iothub_device_update_account.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+* `identity` - (Optional) An `identity` block as defined below.
+
+* `public_network_access` - (Optional) The Public Network Access setting of the IoT Hub Device Update Account. Possible values are `Enabled` and `Disabled`. Defaults to `Enabled`.
```
this is a bool?
```suggestion
* `public_network_access_enabled` - (Optional) The Public Network Access setting of the IoT Hub Device Update Account. Possible values are `Enabled` and `Disabled`. Defaults to `Enabled`.
```

katbyte commented on website/docs/r/iothub_device_update_instance.html.markdown:
```diff
@@ -0,0 +1,108 @@
+
+* `name` - (Required) Specifies the name which should be used for this IoT Hub Device Update Instance. Changing this forces a new resource to be created.
+
+* `iothub_device_update_account_id` - (Required) Specifies the ID of the IoT Hub Device Update Account where the IoT Hub Device Update Instance exists. Changing this forces a new resource to be created.
```
iothub is in the rsource name so
```suggestion
* `device_update_account_id` - (Required) Specifies the ID of the IoT Hub Device Update Account where the IoT Hub Device Update Instance exists. Changing this forces a new resource to be created.
```

myc2h6o commented:
Hi @katbyte thanks for reviewing the pr! I've renamed `iothub_device_update_account_id` to `device_update_account_id` and `public_network_access` (string) to `public_network_access_enabled` (bool). Also simplified some test cases. Please take a look.

katbyte reviewed (approved):
Thanks @myc2h6o - LGTM 🗺️
